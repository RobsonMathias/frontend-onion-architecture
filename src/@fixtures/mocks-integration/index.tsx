export * from './store.mock';
export * from './adapters.mock';
export * from './../mocks/http.mock';
export * from './../mocks/router.mock';
