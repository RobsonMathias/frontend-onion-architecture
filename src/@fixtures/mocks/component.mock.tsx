import React from 'react';

export function Mock() {
  return <div />;
}

export const mockComponent = (componentsName: string[] = []) => {
  let composed: { [key: string]: React.ReactNode } = {};
  componentsName.forEach((name: string) => {
    composed[name] = Mock;
  });
  return composed;
};
