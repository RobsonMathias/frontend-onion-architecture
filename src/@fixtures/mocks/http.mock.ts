export const mockHttp = {
  patch: jest.fn(),
  get: jest.fn(),
  post: jest.fn(),
  put: jest.fn(),
  delete: jest.fn(),
  interceptors: {
    request: {
      use: jest.fn(),
      eject: jest.fn()
    },
    response: {
      use: jest.fn()
    }
  }
} as any;
