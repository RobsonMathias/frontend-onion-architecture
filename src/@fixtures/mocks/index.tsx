export * from './store.mock';
export * from './router.mock';
export * from './component.mock';
export * from './wrappers';
export * from './http.mock';
export * from './services.mock';
export * from './navigation.mock';
export * from './storage.mock';
export * from './use-auth.mock';
