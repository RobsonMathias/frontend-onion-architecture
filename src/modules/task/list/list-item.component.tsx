import React from 'react';
import { Task, TaskStatus } from '@types';
import { Card, CardContent, Grid, IconButton, MenuItem, Select, Typography } from '@mui/material';
import { BiTrash } from 'react-icons/bi';

export enum ListEvent {
  REMOVE = 'remove',
  EDIT = 'edit',
  CHANGE_STATUS = 'status'
}

export type ListItemComponentProps = {
  task: Task;
  onEvent: (task: Task, event: ListEvent) => void;
};

export function ListItemComponent(props: ListItemComponentProps) {
  const { task, onEvent } = props;

  function handleRemove() {
    onEvent(task, ListEvent.REMOVE);
  }

  function handleEdit() {
    onEvent(task, ListEvent.EDIT);
  }

  function handleStatus(evt: any) {
    onEvent({ ...task, status: evt.target.value }, ListEvent.CHANGE_STATUS);
  }

  return (
    <Card>
      <CardContent>
        <Grid container justifyContent="space-between" alignItems="center">
          <Typography
            variant="subtitle1"
            sx={{ cursor: 'pointer', flexGrow: 1 }}
            data-testid="edit"
            onClick={handleEdit}
          >
            {task.description}
          </Typography>
          <Select
            labelId={`task-status-label-${task.id}`}
            id={`task-status-${task.id}`}
            inputProps={{ ['data-testid']: 'status' }}
            value={task.status}
            label="Status"
            onChange={handleStatus}
          >
            <MenuItem value={TaskStatus.NEW}>Novo</MenuItem>
            <MenuItem value={TaskStatus.PROGRESS}>Trabalhando</MenuItem>
            <MenuItem value={TaskStatus.DONE}>Finalizado</MenuItem>
          </Select>
          <IconButton onClick={handleRemove} aria-label="Remover" data-testid="remove">
            <BiTrash />
          </IconButton>
        </Grid>
      </CardContent>
    </Card>
  );
}
