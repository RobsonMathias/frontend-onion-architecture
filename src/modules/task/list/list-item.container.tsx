import React, { useEffect, useState } from 'react';

import { ListItemComponent, ListEvent } from './list-item.component';
import { Task } from '@types';
import { EditContainer } from '../edit';
import { useTask } from '../hooks';

export type ListItemContainerProps = {
  task: Task;
};

export function ListItemContainer(props: ListItemContainerProps) {
  const { task } = props;
  const { remove, setStatus } = useTask();
  const [editTask, setEditTask] = useState(false);

  function handleEvent(task: Task, event: ListEvent) {
    switch (event) {
      case ListEvent.CHANGE_STATUS:
        setStatus(task, task.status);
        break;
      case ListEvent.EDIT:
        setEditTask(true);
        break;
      case ListEvent.REMOVE:
        remove(task.id).then();
        break;
    }
  }

  function handleComplete() {
    setEditTask(false);
  }

  return (
    <>
      {editTask ? (
        <EditContainer task={task} onCompleted={handleComplete} />
      ) : (
        <ListItemComponent task={task} onEvent={handleEvent} />
      )}
    </>
  );
}
