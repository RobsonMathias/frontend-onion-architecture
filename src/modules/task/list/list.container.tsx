import React, { useEffect } from 'react';

import { useList } from '../hooks';
import { LoadingComponent } from './loading.component';
import { STORE_STATUS_FINISHED } from 'constants/store';
import { Box } from '@mui/material';
import { ListItemContainer } from './list-item.container';

export function ListContainer() {
  const { list, tasks } = useList();

  useEffect(() => {
    list();
  }, []);

  return tasks.status === STORE_STATUS_FINISHED ? (
    <>
      {tasks.data.map((task) => (
        <Box key={task.id} mb={1}>
          <ListItemContainer task={task} />
        </Box>
      ))}
    </>
  ) : (
    <LoadingComponent />
  );
}
