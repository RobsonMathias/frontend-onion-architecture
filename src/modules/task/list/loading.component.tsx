import React from 'react';
import { Card, CardContent, Grid, Skeleton } from '@mui/material';

export function LoadingComponent() {
  return (
    <>
      {Array.from({ length: 10 }, (v, k) => k).map((i) => (
        <Card key={i}>
          <CardContent>
            <Grid container justifyContent="space-between">
              <Skeleton width={300} />
              <Skeleton width={40} />
            </Grid>
          </CardContent>
        </Card>
      ))}
    </>
  );
}
