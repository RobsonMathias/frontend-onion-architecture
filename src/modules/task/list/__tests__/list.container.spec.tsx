/**
 * @jest-environment jsdom
 */
import '@testing-library/jest-dom/extend-expect';

import { render } from '@testing-library/react';
import React from 'react';
import {
  STORE_TASK_FETCH_LIST_MOCK_FINISHED,
  STORE_TASK_FETCH_LIST_MOCK_LOADING
} from '@fixtures/stores/task';
import { mockComponent } from '@fixtures/mocks';

const mockUseList = jest.fn();
const mockList = jest.fn();
jest.mock('../../hooks', () => ({
  useList: jest.fn((...args) => mockUseList(...args))
}));
jest.mock('./../list-item.container', () => mockComponent(['ListItemContainer']));

import { ListContainer } from '../list.container';

describe('ListContainer', () => {
  it('should render successfully', () => {
    mockUseList.mockImplementation(() => ({
      list: mockList,
      tasks: STORE_TASK_FETCH_LIST_MOCK_FINISHED
    }));
    const { container } = render(<ListContainer />);
    expect(container).toMatchSnapshot();
  });

  it('should render loading successfully', () => {
    mockUseList.mockImplementation(() => ({
      list: mockList,
      tasks: STORE_TASK_FETCH_LIST_MOCK_LOADING
    }));
    const { container } = render(<ListContainer />);
    expect(container).toMatchSnapshot();
  });
});
