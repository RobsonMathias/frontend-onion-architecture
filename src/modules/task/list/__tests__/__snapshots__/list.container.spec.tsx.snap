// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListContainer should render loading successfully 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container css-1xzqqmo-MuiGrid-root"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          style="width: 300px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          style="width: 40px;"
        />
      </div>
    </div>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container css-1xzqqmo-MuiGrid-root"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          style="width: 300px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          style="width: 40px;"
        />
      </div>
    </div>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container css-1xzqqmo-MuiGrid-root"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          style="width: 300px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          style="width: 40px;"
        />
      </div>
    </div>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container css-1xzqqmo-MuiGrid-root"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          style="width: 300px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          style="width: 40px;"
        />
      </div>
    </div>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container css-1xzqqmo-MuiGrid-root"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          style="width: 300px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          style="width: 40px;"
        />
      </div>
    </div>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container css-1xzqqmo-MuiGrid-root"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          style="width: 300px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          style="width: 40px;"
        />
      </div>
    </div>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container css-1xzqqmo-MuiGrid-root"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          style="width: 300px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          style="width: 40px;"
        />
      </div>
    </div>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container css-1xzqqmo-MuiGrid-root"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          style="width: 300px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          style="width: 40px;"
        />
      </div>
    </div>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container css-1xzqqmo-MuiGrid-root"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          style="width: 300px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          style="width: 40px;"
        />
      </div>
    </div>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container css-1xzqqmo-MuiGrid-root"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          style="width: 300px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          style="width: 40px;"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`ListContainer should render successfully 1`] = `
<div>
  <div
    class="MuiBox-root css-19idom"
  >
    <div />
  </div>
</div>
`;
